core/src/main/java/nu/yona/server/analysis/service/ActivityService.java

```diff
@@ -30,48 +40,172 @@
+	{
+		if (!containsActivityForGoal(activityEntitiesAtDate, activeGoal))
+		{
+			activityEntitiesAtDate.add(zeroActivityEntitySupplier.apply(activeGoal, dateAtStartOfDay));
+		}
+	}
+
+	private <T extends IntervalActivity> boolean containsActivityForGoal(Set<T> dayActivityEntitiesAtDate, Goal goal)
+	{
+		return dayActivityEntitiesAtDate.stream().anyMatch(a -> a.getGoal().equals(goal));
+	}
+
 	public WeekActivityDTO getWeekActivityDetail(UUID userID, LocalDate date, UUID goalID)
 	{
-		UUID userAnonymizedID = userService.getPrivateUser(userID).getPrivateData().getUserAnonymizedID();
-		return WeekActivityDTO.createInstance(WeekActivity.getRepository().findOne(userAnonymizedID, date, goalID),
-				LevelOfDetail.WeekDetail);
+		UUID userAnonymizedID = userService.getUserAnonymizedID(userID);
+		WeekActivity weekActivityEntity = weekActivityRepository.findOne(userAnonymizedID, date, goalID);
+		if (weekActivityEntity == null)
+		{
+			weekActivityEntity = getMissingZeroActivity(userID, date, goalID, userAnonymizedID, ChronoUnit.WEEKS,
+					(goal, startOfWeek) -> WeekActivity.createInstanceZeroActivity(null, goal, startOfWeek));
+		}
+		return WeekActivityDTO.createInstance(weekActivityEntity, LevelOfDetail.WeekDetail);
 	}
 
 	public DayActivityDTO getDayActivityDetail(UUID userID, LocalDate date, UUID goalID)
 	{
-		UUID userAnonymizedID = userService.getPrivateUser(userID).getPrivateData().getUserAnonymizedID();
-		return DayActivityDTO.createInstance(DayActivity.getRepository().findOne(userAnonymizedID, date, goalID),
-				LevelOfDetail.DayDetail);
+		UUID userAnonymizedID = userService.getUserAnonymizedID(userID);
+		DayActivity dayActivityEntity = dayActivityRepository.findOne(userAnonymizedID, date, goalID);
+		if (dayActivityEntity == null)
+		{
+			dayActivityEntity = getMissingZeroActivity(userID, date, goalID, userAnonymizedID, ChronoUnit.DAYS,
+					(goal, startOfDay) -> DayActivity.createInstanceZeroActivity(null, goal, startOfDay));
+		}
+		return DayActivityDTO.createInstance(dayActivityEntity, LevelOfDetail.DayDetail);
+	}
+
+	private <T extends IntervalActivity> T getMissingZeroActivity(UUID userID, LocalDate date, UUID goalID, UUID userAnonymizedID,
+			ChronoUnit timeUnit, BiFunction<Goal, ZonedDateTime, T> zeroActivityEntitySupplier)
+	{
+		UserAnonymizedDTO userAnonymized = userAnonymizedService.getUserAnonymized(userAnonymizedID);
+		Optional<Goal> goal = userAnonymized.getGoals().stream().filter(g -> g.getID().equals(goalID)).findAny();
+		if (!goal.isPresent())
+		{
+			throw GoalServiceException.goalNotFoundById(userID, goalID);
+		}
+		ZonedDateTime dateAtStartOfDay = date.atStartOfDay(ZoneId.of(userAnonymized.getTimeZoneId()));
+		if (!goal.get().getCreationTime().isBefore(dateAtStartOfDay.plus(1, timeUnit)))
+		{
+			throw ActivityServiceException.activityDateGoalMismatch(userID, date, goalID);
+		}
+		return zeroActivityEntitySupplier.apply(goal.get(), dateAtStartOfDay);
+	}
+
+	class StartEndDate
```

Comment:
No problem, Interval sounds OK to me.

On Wed, Apr 13, 2016 at 8:15 PM, Bert Roos <notifications@github.com> wrote:

> In core/src/main/java/nu/yona/server/analysis/service/ActivityService.java
> <https://github.com/yonadev/yona-server/pull/149#discussion_r59597932>:
>
> > +	{
> > +		UserAnonymizedDTO userAnonymized = userAnonymizedService.getUserAnonymized(userAnonymizedID);
> > +		Optional<Goal> goal = userAnonymized.getGoals().stream().filter(g -> g.getID().equals(goalID)).findAny();
> > +		if (!goal.isPresent())
> > +		{
> > +			throw GoalServiceException.goalNotFoundById(userID, goalID);
> > +		}
> > +		ZonedDateTime dateAtStartOfDay = date.atStartOfDay(ZoneId.of(userAnonymized.getTimeZoneId()));
> > +		if (!goal.get().getCreationTime().isBefore(dateAtStartOfDay.plus(1, timeUnit)))
> > +		{
> > +			throw ActivityServiceException.activityDateGoalMismatch(userID, date, goalID);
> > +		}
> > +		return zeroActivityEntitySupplier.apply(goal.get(), dateAtStartOfDay);
> > +	}
> > +
> > +	class StartEndDate
>
> Yes, you're right.
> But still... StartEndDate is a very confusing thing. If you have anything
> better than Interval, then please suggest.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/yonadev/yona-server/pull/149/files/be59c945064070e3ed1326e86d8268355c240fd7#r59597932>
>